lib/tapioca/compilers/dsl/active_record_relations.rb

```diff
@@ -0,0 +1,465 @@
+# typed: strict
+# frozen_string_literal: true
+begin
+  require "active_record"
+rescue LoadError
+  return
+end
+
+require "tapioca/compilers/dsl/helper/active_record_constants"
+
+module Tapioca
+  module Compilers
+    module Dsl
+      class ActiveRecordRelations < Base
+        extend T::Sig
+
+        sig do
+          override
+            .params(root: RBI::Tree, constant: T.class_of(::ActiveRecord::Base))
+            .void
+        end
+        def decorate(root, constant)
+          root.create_path(constant) do |model|
+            constant_name = T.must(qualified_name_of(constant))
+            RelationGenerator.new(model, constant_name).generate
+          end
+        end
+
+        sig { override.returns(T::Enumerable[Module]) }
+        def gather_constants
+          ActiveRecord::Base.descendants.reject(&:abstract_class?)
+        end
+
+        class RelationGenerator
+          extend T::Sig
+          include ParamHelper
+          include Reflection
+          include Helper::ActiveRecordConstants
+
+          sig do
+            params(
+              model: RBI::Scope,
+              constant_name: String
+            ).void
+          end
+          def initialize(model, constant_name)
+            @model = model
+            @constant_name = constant_name
+            @relation_methods_module = T.let(
+              model.create_module(RelationMethodsModuleName),
+              RBI::Scope
+            )
+            @association_relation_methods_module = T.let(
+              model.create_module(AssociationRelationMethodsModuleName),
+              RBI::Scope
+            )
+            @common_relation_methods_module = T.let(
+              model.create_module(CommonRelationMethodsModuleName),
+              RBI::Scope
+            )
+          end
+
+          sig { void }
+          def generate
+            create_classes_and_includes
+            create_common_methods
+            create_relation_methods
+            create_association_relation_methods
+          end
+
+          ASSOCIATION_METHODS = T.let(
+            ::ActiveRecord::AssociationRelation.instance_methods -
+              ::ActiveRecord::Relation.instance_methods,
+            T::Array[Symbol]
+          )
+          COLLECTION_PROXY_METHODS = T.let(
+            ::ActiveRecord::Associations::CollectionProxy.instance_methods -
+              ::ActiveRecord::AssociationRelation.instance_methods,
+            T::Array[Symbol]
+          )
+
+          QUERY_METHODS = T.let(begin
+            # Grab all Query methods
+            query_methods = ActiveRecord::QueryMethods.instance_methods(false)
+            # Grab all Spawn methods
+            query_methods |= ActiveRecord::SpawnMethods.instance_methods(false)
+            # Remove the ones we know are private API
+            query_methods -= [:arel, :build_subquery, :construct_join_dependency, :extensions, :spawn]
```

Comment:
Should we move this a `PRIVATE_API_METHODS` constant?

Reply:
I was trying to keep all the constant definitions as "lists of methods we want" and that's why I didn't pull this out to a constant. But, if you think it makes sense, I can do it.